### ISSUE TITLE
BOOST 1_65

### ISSUE THREAD

**Issue by artemp** [mapnik/mapnik#3793]
https://stackoverflow.com/questions/41770375/strange-semantic-behaviour-of-boost-spirit-x3-after-splitting/41785268#41785268

Master DEBUG=y builds are affected by workaround introduced in boost-1_65
/cc @springmeyer

**Comment by artemp** [comment]
hmm tricky..

**Comment by springmeyer** [comment]
@artemp thanks for creating this ticket. Am I understanding right that boost 1.65.1 is now working in mapnik master, but not in debug mode? Does 3.0.x/ v3.0.16 also suffer from boost 1.65.1 issue, or just master?

**Comment by springmeyer** [comment]
> Am I understanding right that boost 1.65.1 is now working in mapnik master, but not in debug mode? 

Seeing the crash in https://github.com/mapnik/mapnik/tree/boost-1.65.1 at https://travis-ci.org/mapnik/mapnik/builds/303149182.

```
run: mason_packages/.link/include/boost/spirit/home/x3/nonterminal/rule.hpp:137: std::string boost::spirit::x3::get_info<boost::spirit::x3::rule<mapnik::json::grammar::unicode_string_tag, std::basic_string<char>, false>, void>::operator()(const T &) const [Parser = boost::spirit::x3::rule<mapnik::json::grammar::unicode_string_tag, std::basic_string<char>, false>, Enable = void]: Assertion `(r.name)&&("uninitialized rule")' failed.
./test/run: line 60: 11016 Aborted                 (core dumped) ./test/unit/run
[1m[36m* Running standalone C++ tests...[0m
agg_rasterizer_integer_overflow_test-bin: mason_packages/.link/include/boost/spirit/home/x3/nonterminal/rule.hpp:137: std::string boost::spirit::x3::get_info<boost::spirit::x3::rule<mapnik::json::grammar::positions_tag, mapbox::util::variant<mapbox::geometry::point<double>, std::vector<mapbox::geometry::point<double>, std::allocator<mapbox::geometry::point<double> > >, std::vector<std::vector<mapbox::geometry::point<double>, std::allocator<mapbox::geometry::point<double> > >, std::allocator<std::vector<mapbox::geometry::point<double>, std::allocator<mapbox::geometry::point<double> > > > >, std::vector<std::vector<std::vector<mapbox::geometry::point<double>, std::allocator<mapbox::geometry::point<double> > >, std::allocator<std::vector<mapbox::geometry::point<double>, std::allocator<mapbox::geometry::point<double> > > > >, std::allocator<std::vector<std::vector<mapbox::geometry::point<double>, std::allocator<mapbox::geometry::point<double> > >, std::allocator<std::vector<mapbox::geometry::point<double>, std::allocator<mapbox::geometry::point<double> > > > > > > >, false>, void>::operator()(const T &) const [Parser = boost::spirit::x3::rule<mapnik::json::grammar::positions_tag, mapbox::util::variant<mapbox::geometry::point<double>, std::vector<mapbox::geometry::point<double>, std::allocator<mapbox::geometry::point<double> > >, std::vector<std::vector<mapbox::geometry::point<double>, std::allocator<mapbox::geometry::point<double> > >, std::allocator<std::vector<mapbox::geometry::point<double>, std::allocator<mapbox::geometry::point<double> > > > >, std::vector<std::vector<std::vector<mapbox::geometry::point<double>, std::allocator<mapbox::geometry::point<double> > >, std::allocator<std::vector<mapbox::geometry::point<double>, std::allocator<mapbox::geometry::point<double> > > > >, std::allocator<std::vector<std::vector<mapbox::geometry::point<double>, std::allocator<mapbox::geometry::point<double> > >, std::allocator<std::vector<mapbox::geometry::point<double>, std::allocator<mapbox::geometry::point<double> > > > > > > >, false>, Enable = void]: Assertion `(r.name)&&("uninitialized rule")' failed.
./test/run: line 33: 11019 Aborted                 (core dumped) ${FILE}
datasource_registration_test-bin: mason_packages/.link/include/boost/spirit/home/x3/nonterminal/rule.hpp:137: std::string boost::spirit::x3::get_info<boost::spirit::x3::rule<mapnik::json::grammar::positions_tag, mapbox::util::variant<mapbox::geometry::point<double>, std::vector<mapbox::geometry::point<double>, std::allocator<mapbox::geometry::point<double> > >, std::vector<std::vector<mapbox::geometry::point<double>, std::allocator<mapbox::geometry::point<double> > >, std::allocator<std::vector<mapbox::geometry::point<double>, std::allocator<mapbox::geometry::point<double> > > > >, std::vector<std::vector<std::vector<mapbox::geometry::point<double>, std::allocator<mapbox::geometry::point<double> > >, std::allocator<std::vector<mapbox::geometry::point<double>, std::allocator<mapbox::geometry::point<double> > > > >, std::allocator<std::vector<std::vector<mapbox::geometry::point<double>, std::allocator<mapbox::geometry::point<double> > >, std::allocator<std::vector<mapbox::geometry::point<double>, std::allocator<mapbox::geometry::point<double> > > > > > > >, false>, void>::operator()(const T &) const [Parser = boost::spirit::x3::rule<mapnik::json::grammar::positions_tag, mapbox::util::variant<mapbox::geometry::point<double>, std::vector<mapbox::geometry::point<double>, std::allocator<mapbox::geometry::point<double> > >, std::vector<std::vector<mapbox::geometry::point<double>, std::allocator<mapbox::geometry::point<double> > >, std::allocator<std::vector<mapbox::geometry::point<double>, std::allocator<mapbox::geometry::point<double> > > > >, std::vector<std::vector<std::vector<mapbox::geometry::point<double>, std::allocator<mapbox::geometry::point<double> > >, std::allocator<std::vector<mapbox::geometry::point<double>, std::allocator<mapbox::geometry::point<double> > > > >, std::allocator<std::vector<std::vector<mapbox::geometry::point<double>, std::allocator<mapbox::geometry::point<double> > >, std::allocator<std::vector<mapbox::geometry::point<double>, std::allocator<mapbox::geometry::point<double> > > > > > > >, false>, Enable = void]: Assertion `(r.name)&&("uninitialized rule")' failed.

```

**Comment by artemp** [comment]
@springmeyer - yes ^ .  The run-time error is triggered by added in 1.65.x assert.  I spent some time trying to re-arrange grammars to avoid this assert with no success, yet. As far as I can tell this is not causing any logic breakage in our case so it's ok to ignore/disable. How feasible is to patch up one boost header for the coverage build ?

**Comment by springmeyer** [comment]
> The run-time error is triggered by added in 1.65.x assert. I spent some time trying to re-arrange grammars to avoid this assert with no success, yet.

Is there a boost bug for this? Do we need to report it?

> How feasible is to patch up one boost header for the coverage build ?

Feasible. Since these are mason packages being used, we could patch the boost headers in the mason package or we could dynamically find/replace or patch the file on travis, since it will be locally installed in `./mason_packages/headers/boost`.

**Comment by artemp** [comment]
This has been fixed in #3821.
